File: Problem3_ContainerWithMostWater.py
#Time Complexity: O(n)
#Space Complexity: O(1)
#LeetCode: Yes

'''
Approach: Prioritize height instead of width. Height can increase, width will always decrease.
'''
class Solution:
    def maxArea(self, height):
        i, j = 0, len(height)-1
        max_area = 0
        while(i<j):
            max_area = max(max_area, min(height[i], height[j])*(j-i))
            if height[i]<height[j]:
                i+=1
            else:
                j-=1
        return max_area

obj = Solution()
height = [1,8,6,2,5,4,8,3,7]
print(obj.maxArea(height))

File: Problem3_ContainerWithMostWater.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Using two pointer approach, take one point at 0th index and second point at N-1 index
// By calculating the area of the container formed, decide to move the left or right pointer based on min height
// Have a variable to keep track of the max area
// pointer at the min height can be moved until a greater height is found as it can increase the area

public class Problem3_ContainerWithMostWater {
    public int maxArea(int[] height) {
        int n=height.length;
        int l=0, r=n-1, max=0;
        while(l<r){
            int h = Math.min(height[l], height[r]);
            int area = h * (r-l);
            max = Math.max(max, area);
            while(l<r && height[l]<=h) l++;
            while(l<r && height[r]<=h) r--;
        }
        return max;
    }
}
